src/bytes_mut.rs

```diff
@@ -758,6 +758,119 @@ impl BytesMut {
         debug_assert_eq!(self.len, v.len());
     }
 
+    /// Attempts to cheaply reclaim capacity for at least `additional` more bytes to be inserted
+    /// into the given `BytesMut` and returns `true` if it succeeded.
+    ///
+    /// `try_reclaim` returns false under these conditions:
+    ///  - The spare capacity left is less than `additional` bytes AND
+    ///  - The existing allocation cannot be reclaimed cheaply or it was less than
+    ///    `additional` bytes in size
+    ///
+    /// Reclaiming the allocation cheaply is possible if the `BytesMut` is currently empty and
+    /// there are no outstanding references through other `BytesMut`s or `Bytes` which point to the
+    /// same underlying storage.
+    ///
+    ///  A call to `try_reclaim` never copies inside the allocation nor allocates new storage.
```

Comment:
It's probably worth thinking about the behavior when it is non-empty. The `reserve` method will copy data backwards only if the full capacity is at least twice the amount of data being copied (I think). What logic makes sense here?

Reply:
I guess it depends on usage patterns. For my use case, it is likely the buffer is either completely empty or more than one reference to the underlying storage still exists, therefore it doesn't make a big difference.

I guess it could make a bigger difference when you are reading a lot into a large buffer, start reading the second thing which also takes a while, then the first thing is handled and when you call `try_reclaim` you either copy the large thing to the front of the buffer or reclaim tells you "couldn't reclaim", but I am not sure what you would want in that case

Reply:
Thinking about this further, it would mean `try_reclaim` would become more expensive to call, because we wouldn't just check whether the `BytesMut` is empty but also which kind it is and whether we can move bytes, before returning. I wonder if @braddunbar has an opinion?

Reply:
I think it's okay to just always copy the data back. After all, the user knows what the length is and how much additional space they are requesting. If the user doesn't want to pay the cost of copying data back, then the user can call `clear` or just not use `try_reclaim`.